lib/src/config_provider/config.dart

```diff
@@ -215,363 +219,699 @@ class Config {
+                    'Objective C support is EXPERIMENTAL. The API may change '
+                    'in a breaking way without notice.');
+                return Language.objc;
+              } else {
+                return Language.c;
+              }
+            },
+          ),
+          defaultValue: (node) => Language.c,
+          resultOrDefault: (node) => _language = node.value as Language,
+        ),
+        FixedMapEntry(
+            key: strings.headers,
+            required: true,
+            valueConfigSpec: FixedMapConfigSpec<List<String>, Headers>(
+              entries: [
+                FixedMapEntry(
+                  key: strings.entryPoints,
+                  valueConfigSpec: ListSchema<String, List<String>>(
+                      childSchema: StringConfigSpec()),
+                  required: true,
+                ),
+                FixedMapEntry(
+                  key: strings.includeDirectives,
+                  valueConfigSpec: ListSchema<String, List<String>>(
+                      childSchema: StringConfigSpec()),
+                ),
+              ],
+              transform: (node) => headersExtractor(node.value, filename),
+              result: (node) => _headers = node.value,
+            )),
+        FixedMapEntry(
+          key: strings.compilerOpts,
+          valueConfigSpec: OneOfConfigSpec<List<String>, List<String>>(
+            childConfigSpecs: [
+              StringConfigSpec(
+                transform: (node) => [node.value],
+              ),
+              ListSchema<String, List<String>>(childSchema: StringConfigSpec())
+            ],
+            transform: (node) => compilerOptsExtractor(node.value),
+          ),
+          defaultValue: (node) => <String>[],
+          resultOrDefault: (node) => _compilerOpts = node.value as List<String>,
+        ),
+        FixedMapEntry(
+            key: strings.compilerOptsAuto,
+            valueConfigSpec: FixedMapConfigSpec(
+              entries: [
+                FixedMapEntry(
+                  key: strings.macos,
+                  valueConfigSpec: FixedMapConfigSpec(
+                    entries: [
+                      FixedMapEntry(
+                        key: strings.includeCStdLib,
+                        valueConfigSpec: BoolConfigSpec(),
+                        defaultValue: (node) => true,
+                      )
+                    ],
+                  ),
+                )
+              ],
+              transform: (node) => CompilerOptsAuto(
+                macIncludeStdLib: ((node.value)[strings.macos]
+                    as Map?)?[strings.includeCStdLib] as bool,
+              ),
+              result: (node) => _compilerOpts.addAll(
+                  (node.value as CompilerOptsAuto).extractCompilerOpts()),
+            )),
+        FixedMapEntry(
+          key: strings.libraryImports,
+          valueConfigSpec:
+              DynamicMapConfigSpec<String, Map<String, LibraryImport>>(
```

Comment:
Can we merge `DynamicMapConfigSpec` with `FixedMapConfigSpec`? And just have a boolean field and that field in the constructor?

It feels that conceptually these really should have the same API and implementation except for whether the keys are defined.

On difference is that all keys and values in the dynamic one need to have the same type while this is not true for the fixed one. If that is a more fundamental difference, then the two classes should be called `Homogeneous` and `Hetrogeneous` probably. However, that can also be expressed with the type argument for the values being `dynamic` for the hetrogeneous one.

Are there any other reasons why these two types are not merged?

Reply:
One of the reasons they are separate is handling default values via FixedMapEntry.

1. We could merge the two, but then have 2 separate lists? One -`final List<FixedMapEntry> entries;` and another - `final List<({String keyRegexp, ConfigSpec valueConfigSpec})> keyValueConfigSpecs;`
2. Or we could keep one class, but have 2 different types of MapEntry - `FixedMapEntry` and `DynamicMapEntry`. So that we can separate these. Any class with a DynamicMapEntry in list would also automatically have `additionalProperties` set to true.

I like 2 better. 

> then the two classes should be called Homogeneous and Hetrogeneous probably. However, that can also be expressed with the type argument for the values being dynamic for the hetrogeneous one.

We could just rename them, but the type argument being dynamic would remove any validations

Reply:
Okay, maybe we should just keep them separate, even though both of them are Maps.

https://www.google.com/search?q=heterogeneous+map
https://www.google.com/search?q=heterogeneous+map+java

It looks like `HeterogeneousMap` is the right name.

Maybe `HomogenousMap` should just be `Map`, since all maps `<K, V>` are homogeneous already.

(If it makes sense and there's duplicated code, it could be in a private base class of both.)

Reply:
Um, I don't have much opinion over the names, so would the new naming be -
FixedMapConfigSpec -> HeterogeneousMapConfigSpec
DynamicMapConfigSpec -> HomogenousMapConfigSpec
FixedMapEntry -> HeterogeneousMapEntry

Also technically even DynamicMap can have different values (so not exactly homogenous), so basically a key with a particular pattern must have a value that conforms to a specified ConfigSpec

Reply:
`FixedMapConfigSpec` -> `HeterogeneousMapConfigSpec`
`DynamicMapConfigSpec` -> `MapConfigSpec` (because the way the `<K,V>` works here is just how normal maps work)
`FixedMapEntry` -> `HeterogeneousMapEntry`